core/src/main/java/org/apache/carbondata/core/datamap/TableDataMap.java

```diff
@@ -120,37 +132,166 @@ public BlockletDetailsFetcher getBlockletDetailsFetcher() {
    * @param filterExp
    * @return
    */
-  public List<ExtendedBlocklet> prune(List<Segment> segments, FilterResolverIntf filterExp,
-      List<PartitionSpec> partitions) throws IOException {
-    List<ExtendedBlocklet> blocklets = new ArrayList<>();
-    SegmentProperties segmentProperties;
-    Map<Segment, List<DataMap>> dataMaps = dataMapFactory.getDataMaps(segments);
+  public List<ExtendedBlocklet> prune(List<Segment> segments, final FilterResolverIntf filterExp,
+      final List<PartitionSpec> partitions) throws IOException {
+    final List<ExtendedBlocklet> blocklets = new ArrayList<>();
+    final Map<Segment, List<DataMap>> dataMaps = dataMapFactory.getDataMaps(segments);
+    // for non-filter queries
+    if (filterExp == null) {
+      // if filter is not passed, then return all the blocklets.
+      return pruneWithoutFilter(segments, partitions, blocklets);
+    }
+    // for filter queries
+    int totalFiles = 0;
+    boolean isBlockDataMapType = true;
+    for (Segment segment : segments) {
+      for (DataMap dataMap : dataMaps.get(segment)) {
+        if (!(dataMap instanceof BlockDataMap)) {
```

Comment:
This flow can be used by all datamaps why only for blockdatamap?

Reply:
Two reasons:

1.  number of datamaps will be very less if it is not a block or blocklet datamap. Hence multi-threading is not required (as it is overhead for driver in concurrent scenarios)

2. Other datamaps doesn't have number entries count in them.

I will check

Reply:
This one, I have to figure out, number of entries in all kinds of datamap and need to test those scenario. I will handle in next PR